### PR TITLE
Feature/trainings

### PR DESCRIPTION
**Pull request by brauliodiez**
Since this branch has been created starting for students (Students was not 100% ready to merge but it was needed for this case to progress).

Pull request merge updates from trainings API to the students API branch, procedure to follow

- Review Trainings API branch.
- Merge into Students API feature.
- Review Students API + Trainings.
- Merge into master.

### CODE REVIEW

File: src/model/trainer.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export class Trainer{}
```
Comment (Nasdan) [inline src/model/trainer.ts]: Is this model to remove?

File: src/model/training.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import {Trainer} from "./trainer";
+
+export class Training {
+  id : number;
+  name : string;
+  isActive : boolean;
+  start : Date;
+  end : Date;
+  trainers : Array<Trainer>;
+  students : Array<Student>;
```
Comment (brauliodiez) [inline src/model/training.ts]: I like this idea, by doing this we can just get in one step from the API what is needed to edit a given training.

File: src/rest-api/spec/training.spec.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+      // Assert
+      expect(trainingApi.trainingList).to.be.equal(trainingList);
+    });
+  });
+
+  describe('getSummaryTrainingList', () => {
+    it('Get the expected summary training list', sinon.test((done) => {
+
+      trainingApi.setMockDataSeed(trainingList);
```
Comment (brauliodiez) [inline src/rest-api/spec/training.spec.ts]: Missing comment 'Arrange'

File: src/rest-api/spec/training.spec.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+describe('TrainingApi', () => {
+  it('is defined', () => {
+      expect(trainingApi).not.to.be.undefined;
+  })
+
+  describe('setMockDataSeed', () => {
+    it('set the right mockdata seed', () => {
+      // Act
+      trainingApi.setMockDataSeed(trainingList);
```
Comment (Nasdan) [inline src/rest-api/spec/training.spec.ts]: Is this really necessary, expose a public method to set mock data? What is the reason?

Comment (brauliodiez) [inline src/rest-api/spec/training.spec.ts]: The reason behind this, is that we are using a singleton (and the singleton data is feeded in the constructor).

This is just for the mock api, is something temporary, later on we will remove this.

We could add something more complex, like simulating and AJAX call or something like that, but not sure if this applies at this stage.

File: src/model/spec/training.spec.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  describe('#constructor', () => {
+    it('Is initialized with the expected values on properties', () => {
+      expect(training.id).to.be.equal(-1);
+      expect(training.name).to.be.equal('');
+      expect(training.isActive).to.be.false;
+      expect(training.start.toString()).to.be.equal(new Date().toString());
+      expect(training.end.toString()).to.be.equal(new Date().toString());
+      expect(training.trainers.length).to.be.equal(new Array<Trainer>().length);
+      expect(training.students.length).to.be.equal(new Array<Student>().length);
+  });
```
Comment (franmolmedo) [inline src/model/spec/training.spec.ts]: to.be.equal(0) ?

File: src/rest-api/training.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  constructor() {
+    this.trainingList = trainingMockData;
+  }
+
+  setMockDataSeed(trainingList : Training[]) {
+    this.trainingList = trainingList;
+  }
+
+  getSummaryTrainingList() : Promise<TrainingSummary[]> {
+    const trainingSummaryList : TrainingSummary[] = this.trainingList.map((training) => {
```
Comment (franmolmedo) [inline src/rest-api/training.ts]: brackets in this kind of map function's argument are not necessary

Comment (brauliodiez) [inline src/rest-api/training.ts]: Good catch

File: src/model/spec/training.spec.ts (hunk 5 above)
Comment (brauliodiez) [inline src/model/spec/training.spec.ts]: Good idea

File: src/rest-api/spec/training.spec.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        // Assert
+        expect(summaryList).not.to.be.undefined;
+        expect(summaryList.length).to.be.equal(trainingList.length);
+        expect(summaryList[0].id).to.be.equal(trainingList[0].id);
+        expect(summaryList[0].name).to.be.equal(trainingList[0].name);
+        expect(summaryList[0].isActive).to.be.equal(trainingList[0].isActive);
+        expect(summaryList[1].id).to.be.equal(trainingList[1].id);
+        expect(summaryList[1].name).to.be.equal(trainingList[1].name);
+        expect(summaryList[1].isActive).to.be.equal(trainingList[1].isActive);
+
```
Comment (brauliodiez) [inline src/rest-api/spec/training.spec.ts]: Haven't tried that, sounds good, we save code and is clearer for a developer to review it

File: src/model/trainer.ts (hunk 1 above)
Comment (brauliodiez) [inline src/model/trainer.ts]: Nope, we have a main Training entity, then a TrainingSummary (just a map including a summary of the fields).  Training entity will be the one use when we want to edit a given training.

File: src/rest-api/spec/training.spec.ts (hunk 7 above)
Comment (borjamaza) [inline src/rest-api/spec/training.spec.ts]: works fine!

**Comment by brauliodiez** [comment]
Good stuff, goint to merge into students branch case and update issue

**Pull request merged**